src/main/java/org/team4159/frc/robot/subsystems/Drivetrain.java

```diff
@@ -1,36 +1,174 @@
-            (WPI_TalonSRX) configureTalonSRX(new WPI_TalonSRX(CAN_IDS.RIGHT_FRONT_TALON)),
-            (WPI_TalonSRX) configureTalonSRX(new WPI_TalonSRX(CAN_IDS.RIGHT_REAR_TALON)));
+    left_front_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.LEFT_FRONT_FALCON_ID));
+    left_rear_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.LEFT_REAR_FALCON_ID));
+    right_front_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.RIGHT_FRONT_FALCON_ID));
+    right_rear_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.RIGHT_REAR_TALON_ID));
+
+    left_front_falcon.configSelectedFeedbackSensor(FeedbackDevice.IntegratedSensor);
+    right_front_falcon.configSelectedFeedbackSensor(FeedbackDevice.IntegratedSensor);
+
+    left_falcons = new SpeedControllerGroup(
+            (WPI_TalonFX) left_front_falcon,
+            (WPI_TalonFX) left_rear_falcon);
+    right_falcons = new SpeedControllerGroup(
+            (WPI_TalonFX) right_front_falcon,
+            (WPI_TalonFX) right_rear_falcon);
+
+    left_falcons.setInverted(true);
+
+    pigeon = new PigeonIMU(CAN_IDS.PIGEON_ID);
+
+    differential_drive = new DifferentialDrive(left_falcons, right_falcons);
+    odometry = new Odometry(new Pose2d(0.0, 0.0, Rotation2d.fromDegrees(getDirection())));
+
+    resetDirection();
+    resetEncoders();
+  }
+
+  public void flipOrientation() {
+    is_oriented_forward = !is_oriented_forward;
   }
 
-  private TalonSRX configureTalonSRX(TalonSRX talonSRX) {
+  @Override
+  public void periodic() {
+    odometry.update(
+      getLeftDistance(),
+      getRightDistance(),
+      getDirection());
+  }
+
+  private TalonFX configureTalonFX(TalonFX talonSRX) {
     talonSRX.configFactoryDefault();
     talonSRX.setNeutralMode(NeutralMode.Coast);
 
     return talonSRX;
   }
 
-  public void setRawSpeeds(double left, double right) {
-    left_talons.set(left);
-    right_talons.set(right);
+  public void tankDrive(double left, double right) {
+    if (is_oriented_forward) {
+      differential_drive.tankDrive(left, right);
+    } else {
+      differential_drive.tankDrive(-right, -left);
+    }
+  }
+
+  public void arcadeDrive(double forward, double rotation) {
+    if (is_oriented_forward) {
+      differential_drive.arcadeDrive(forward, rotation);
+    } else {
+      differential_drive.arcadeDrive(-forward, -rotation);
+    }
+  }
+
+  public void rawDrive(double left, double right) {
+    if (is_oriented_forward) {
+      left_falcons.set(left);
+      right_falcons.set(right);
+    } else {
+      left_falcons.set(-right);
+      right_falcons.set(-left);
+    }
+  }
+
+  public void voltsDrive(double left_volts, double right_volts) {
+    if (is_oriented_forward) {
+      left_falcons.setVoltage(left_volts);
+      right_falcons.setVoltage(right_volts);
+    } else {
+      left_falcons.setVoltage(-right_volts);
+      right_falcons.setVoltage(-left_volts);
+    }
+  }
+
+  public void setPose(Translation2d coords) {
+    resetEncoders();
```

Comment:
i think resetting encoders will mess stuff up because doesn't odometry still remember the prev_magnitude? we should probably have a way to reset that